### PR TITLE
Add an optional ACME certificate provider.

### PR DESCRIPTION
**Pull request by jmalloc**
This PR adds support for ACME-based certificate generation.

This required a fairly severe overhaul of the certificate "provider" system, which included consolidating the per-provider caches into a single global certificate cache.

There are now two "chains" of certificate providers, one for server names that are "recognized", that is, they are mapped to some known Docker service, and another for those that are "unrecognized".

When a certificate is obtained, it is given a rank based on the provider that was used to obtain it. The providers for "unrecognized" server names are always ranked worse (numerically higher).

When a certificate is requested from the cache, a "worst acceptable rank" value is provided. If the cache certificate has a worse rank than this value, the cache entry is ignored.

The upshot of all of this is that:

- certificates from on-disk files, or from ACME take precedence over generated certs from the Honeycomb CA
- certificates that are obtained for an "unrecognized" server name are discarded if that server name later becomes "recognized" - This can happen if you hit a newly deployed service's public domain before Honeycomb detects the service's existence

By default, the provider "chains" for both recognized and unrecognized server names consists of the `FileProvider` and the `AdhocProvider`.

If the environment variable `ACME_EMAIL` is defined, the `ACMEProvider` is inserted between these two existing providers for the recognized chain.

The `ACME_CACHE` environment variable is the path to directory used to store the certificates obtained via ACME.  According to the `autocert` docs, this can be shared by multiple managers, which I hope/take to mean that multiple Honeycomb instances can share the cache via Docker volumes.

Finally the `ACME_URL` environment variable sets the endpoint for the ACME API.  If this value is undefined it defaults to the Let's Encrypt production API.

### CODE REVIEW

**Comment by jmalloc** [comment]
As far as I know, right at this moment, the code is "complete", but I will need to do further testing in the company sandbox next week.

File: src/cmd/honeycomb/certs.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		config.Certificates.BasePath,
+	)
+
+	return &cert.FileProvider{
+		BasePath: config.Certificates.BasePath,
+		Logger:   logger,
+	}, true
+}
+
+// fileCertificateProvider returns a cert provider that obtains certificates
```
Comment (ezzatron) [inline src/cmd/honeycomb/certs.go]: Should be `acmeCertificateProvider`?

**Review by ezzatron — COMMENTED** [review]
I only gave it a quick scan, but seems logical 👍

**Comment by koshatul** [comment]
I think I've said it before, I'm a huge fan of the `.next()` style of ordering processors instead of using a slice of them.

Other than that I'd need to sit down and read through a lot of changes :)

**Comment by jmalloc** [comment]
Yeah, it's pretty involved.  And I can say right now, that it doesn't work :P

**Comment by jmalloc** [comment]
> I think I've said it before, I'm a huge fan of the `.next()` style of ordering processors instead of using a slice of them.

I use that approach when implementing middlewares - things that can conditionally forward onto the "next stage".

In this case, it's a priority ordered list, it doesn't make sense for each implementation to know about the others. They can also appear in multiple lists with different priorities, meaning conceptually there are different "next" providers to try depending on other factors (whether or not the domain is recognised).

File: src/frontend/cert/file_provider.go [hunk 2]
```diff
@@ -64,12 +63,9 @@ func (p *FileProvider) GetCertificate(
 
 		if ok {
 			p.Logger.Printf(
-				"Loaded certificate for '%s' from '%s', expires at %s (%s), issued by '%s'",
-				n.Unicode,
+				"Loaded certificate %s from from '%s'",
```
Comment (koshatul) [inline src/frontend/cert/file_provider.go:66]: `from from`

**Review by koshatul — CHANGES_REQUESTED** [review]
just the `from from`, everything else looks OK, I might need to check out the branch to actually follow the new code though.

**Pull request closed without merging**